feat: initial implementation

TestingPlant opened the pull request:
Implements #1

TestingPlant committed:
fix: remove mcp_macros

TestingPlant committed:
fix: suppress warning in generated code

TestingPlant committed:
fix: shorten required handler lifetime

TestingPlant committed:
fix: use AsyncFn in HandlerFn

TestingPlant committed:
wip: implement BasicService

The original Service trait will still exist for more advanced users
to implement their own functions.

TestingPlant committed:
feat: implement more functions in BasicService

andrewgazelka commented:
@TestingPlant any updates?

TestingPlant committed:
refactor: create a generic HandlerRegistry

This allows us to maintain a separate tool registry and prompt
registry. Both will have a similar API.

TestingPlant committed:
feat: add prompt registry

TestingPlant committed:
feat: implement ping

TestingPlant committed:
refactor!: switch to the mcp-schema crate

This has a few advantages over the previous automatically generated
schema:

- The ClientRequest enum is tagged, meaning that we can rely on that to
  differentiate between different client request methods without
  hard-coding each method into the code

- Avoids duplicate structs/enums, reducing code complexity

- Reduces compile times by avoiding the need for a build script

- Other crates can use the same protocol types through the mcp-schema
  crate without adding this crate as a dependency

TestingPlant committed:
feat!: connect Service to rpc

Allows the rpc code to process messages using the Service trait. This
commit also adds additional Send/Sync bounds that were previously
missing.

TestingPlant committed:
refactor: move main.rs to separate examples

TestingPlant committed:
refactor!: clone State and relax lifetime bounds

axum already clones the state, so the additional complexity of passing
the state by reference to handlers does not provide any benefits.

TestingPlant committed:
refactor! accept Fn instead of AsyncFn

This adds additional flexibility because the underlying async
function/closure doesn't need to outlive the returned future. In
addition, it avoids the need for nightly features to ensure that the
future is Send.

TestingPlant committed:
refactor: wrap registries in Mutex

This required some lifetime changes as the returned Futures may outlive
the registry.

TestingPlant committed:
fix: type of TextContent is text

TestingPlant committed:
refactor(tool)!: remove required_args

TestingPlant committed:
feat: broadcast errors

TestingPlant committed:
feat: add weather tool

TestingPlant committed:
refactor!: accept PromptContent from tool handler

This allows tool handlers to send images and embedded resources insteaad
of only text.

TestingPlant committed:
fix: clean up warnings

TestingPlant committed:
feat: support client request cancellation

TestingPlant committed:
refactor(prompt): add prompt builder

TestingPlant committed:
feat: implement resources

TestingPlant committed:
feat(weather): add prompts and resources

TestingPlant committed:
refactor: fix clippy lints

TestingPlant commented:
@andrewgazelka The core features have been implemented now. It currently supports: 
- calling tools, prompts, and resources (although template resource URIs are still a work in progress)
- resource subscriptions to notify clients when a resource changes
- request cancellation

Check https://github.com/TestingPlant/mcp/blob/design/mcp/examples/weather.rs for how the new api works in practice. What do you think about it?

TestingPlant committed:
feat: custom service name and version

TestingPlant committed:
refactor(resource)!: split uri into two types

This helps prevent possible runtime errors by using the type system.

andrewgazelka commented:
ok is there a reason the example returns

```
Result<Vec<mcp_schema::PromptMessage>, mcp::Error> {
```

ideally it would return some generic T that can be serialized and has a schema. (should be simple like python ver)

TestingPlant commented:
> ok is there a reason the example returns
> 
> ```
> Result<Vec<mcp_schema::PromptMessage>, mcp::Error> {
> ```
> 
> ideally it would return some generic T that can be serialized and has a schema. (should be simple like python ver)

The prompt message is a string passed directly to the LLM. I believe automatically serializing the output would give the illusion that JSON output is handled in a special manner by the LLM when they're all just interpreted as a string, and there's no output schema in the mcp protocol. Also, this allows other prompt messages, like images and resources, to be returned.

This is also what the python version does: https://github.com/modelcontextprotocol/python-sdk/blob/main/examples/servers/simple-tool/mcp_simple_tool/server.py
`Vec<mcp_schema::PromptMessage>` is the Rust equivalent to the `list[types.TextContent | types.ImageContent | types.EmbeddedResource]` used in the Python version.

andrewgazelka committed:
add rust-toolchain.toml

andrewgazelka commented:
ok will merge in if you add an example with stdio transport as well

andrewgazelka commented:
@TestingPlant

TestingPlant committed:
feat: add stdio transport

TestingPlant committed:
fix: unwrap serve result

TestingPlant commented:
@andrewgazelka stdio transport should work now

andrewgazelka committed:
clean up somewhat

andrewgazelka committed:
update api

andrewgazelka commented:
> @andrewgazelka stdio transport should work now

ok sorry I looked further and I updated the API a bit to be similar to [axum](https://docs.rs/axum/latest/axum/)

https://github.com/ghbountybot/mcp/blob/295df1399758a6fcde735a541cb9e6e8d7a014c4/mcp/examples/weather.rs#L118-L172

However, some things need to be fixed. There were a lot of mutexes that did not need to be there imo. (However, some might still be needed)

Once you get this to pass with this current API I think we are good for merge. @TestingPlant

TestingPlant committed:
fix: fix errors and warnings

TestingPlant committed:
fix: unsubscribe from resources properly

TestingPlant commented:
@andrewgazelka Thanks for helping remove some of the mutexes. I've fixed the errors so the code should work now.

andrewgazelka commented:
ok I am testing @TestingPlant 

<img width="999" alt="image" src="https://github.com/user-attachments/assets/632eadb4-3629-4620-8d82-ff778316a909" />

seems cursor is not picking it up https://www.cursor.com/ .  I am using settings

```
{
    "mcpServers": {
      "server-name": {
        "command": "/Users/andrewgazelka/Projects/bounty/mcp/target/debug/examples/weather",
        "args": []
      }
    }
  }
```

TestingPlant committed:
fix: remove unused deps

TestingPlant committed:
fix: switch to mcp-schema fork

This is a temporary fix while waiting for
https://github.com/yonaka15/mcp-schema/pull/2 to be merged.

TestingPlant commented:
@andrewgazelka I've found the issue in mcp-schema. I'm trying to [get it merged upstream](https://github.com/yonaka15/mcp-schema/pull/2), but I've added 450fad6 to help test it locally.
I would also recommend using [mcp inspector](https://github.com/modelcontextprotocol/inspector) to test it. Cursor has a lot of unimplemented mcp features right now such as resources and prompts.

I'm also confused about the changes to MemoryResource in 24b3f94 - by using a channel like this, `MemoryResource::get` will always panic because `subscribe` will create a receiver that only receives values that were sent after it was subscribed, so immediately calling `try_recv` will almost certainly fail. In addition, `MemoryResource::set` will panic if there are no receivers (typically when the resource exists without being subscribed to). Ignoring the error from `self.tx.send` just means that updates to the resource will be ignored if there are no receivers. What do you think about reverting `MemoryResource` back to the previous version and storing `contents: Arc<Mutex<Vec<mcp_schema::ResourceContents>>>`?

andrewgazelka commented:
> @andrewgazelka I've found the issue in mcp-schema. I'm trying to [get it merged upstream](https://github.com/yonaka15/mcp-schema/pull/2), but I've added [450fad6](https://github.com/ghbountybot/mcp/commit/450fad619e4c30e99dfeb04dc0160353412ae1af) to help test it locally. I would also recommend using [mcp inspector](https://github.com/modelcontextprotocol/inspector?rgh-link-date=2025-03-21T03%3A20%3A35Z) to test it. Cursor has a lot of unimplemented mcp features right now such as resources and prompts.
> 
> I'm also confused about the changes to MemoryResource in [24b3f94](https://github.com/ghbountybot/mcp/commit/24b3f94d28df5aba54b6946b37bd17a6bdff823a) - by using a channel like this, `MemoryResource::get` will always panic because `subscribe` will create a receiver that only receives values that were sent after it was subscribed, so immediately calling `try_recv` will almost certainly fail. In addition, `MemoryResource::set` will panic if there are no receivers (typically when the resource exists without being subscribed to). Ignoring the error from `self.tx.send` just means that updates to the resource will be ignored if there are no receivers. What do you think about reverting `MemoryResource` back to the previous version and storing `contents: Arc<Mutex<Vec<mcp_schema::ResourceContents>>>`?

I had the full mcp schema in `main`

https://github.com/ghbountybot/mcp/blob/main/mcp/src/schema.rs#L2

I do not know why we need to rely on a third party lib. This is generated absed off of Anthropic's official `schema.json` https://github.com/ghbountybot/mcp/blob/main/mcp/schema.json if _this_ is outdated you can just update JSON. Your IDE should have auto completion of types (at least IntelliJ does)

I am fine with impl `MemoryResource` however... I just don't want it to have to be used from a `Mutex`... I think it should have a Mutex internally although I am not 100% set on this. Thoughts?

> Cursor has a lot of unimplemented mcp features right now such as resources and prompts.

kinda confused haven't we implemented resources and prompts? you mean cursor _doesn't_ have them implemented right? 

At end of day I just want this to work with either Windsurf or Cursor (it should be able to be used in production)

TestingPlant commented:
> I do not know why we need to rely on a third party lib. 

I elaborated more on this point in the commit message in ea6bd44. The main advantage is that the ClientRequest enum is tagged, meaning it can determine what type of request it is using the method field. With the automatically generated schema from typify, the request is untagged so we would need to hard code the method strings for each request and check it ourselves. This isn't an issue in the json, it's just a limitation in typify. I believe using the mcp-schema crate is nicer because it helps avoid any future typify limitations that we might encounter. This is a really minor issue though and I could revert it back to the automatically generated schema and add the manual method checking if you'd like.

> you mean cursor doesn't have them implemented right?

Yes. Cursor will say that the service doesn't have any resources and it doesn't appear to use mcp prompts at all. [Resources support seems likely to arrive soon though](https://forum.cursor.com/t/mcp-resources-not-working-in-0-46-9/59960) Tools work though, the only issue blocking tools from working right now is that our resources impl is panicking in the tool handler.

> I think it should have a Mutex internally although I am not 100% set on this. Thoughts?

I believe this would be a good idea, I'll implement it.

TestingPlant committed:
refactor: remove ResourceSlice

Using an Arc to store the resource contents adds additional complexity.
The resource contents must be converted to a Vec to be stored in the
ReadResourcesResult, so using Arc will not reduce allocations.

TestingPlant committed:
fix: store contents in MemoryResource

andrewgazelka commented:
> > I do not know why we need to rely on a third party lib.
> 
> I elaborated more on this point in the commit message in [ea6bd44](https://github.com/ghbountybot/mcp/commit/ea6bd444dd3005d2057103c68aa1579ec0211bb7). The main advantage is that the ClientRequest enum is tagged, meaning it can determine what type of request it is using the method field. With the automatically generated schema from typify, the request is untagged so we would need to hard code the method strings for each request and check it ourselves. This isn't an issue in the json, it's just a limitation in typify. I believe using the mcp-schema crate is nicer because it helps avoid any future typify limitations that we might encounter. This is a really minor issue though and I could revert it back to the automatically generated schema and add the manual method checking if you'd like.
> 
> > you mean cursor doesn't have them implemented right?
> 
> Yes. Cursor will say that the service doesn't have any resources and it doesn't appear to use mcp prompts at all. [Resources support seems likely to arrive soon though](https://forum.cursor.com/t/mcp-resources-not-working-in-0-46-9/59960) Tools work though, the only issue blocking tools from working right now is that our resources impl is panicking in the tool handler.
> 
> > I think it should have a Mutex internally although I am not 100% set on this. Thoughts?
> 
> I believe this would be a good idea, I'll implement it.

ok if you can send a video of it working in mcp inspector with the command will merge. I am having issues with it timing out and such.

TestingPlant commented:
(note: the error output is showing what's in stderr, which is just the debug logging from tracing) 

https://github.com/user-attachments/assets/6822f300-2e84-43c0-a49a-447a549b2b77


https://github.com/user-attachments/assets/0276b058-37fa-40b1-aff7-bcc33f4bbea5

andrewgazelka commented:
merged
